From 42a2f52288d6fd1bbcba390d38d7eb0aae19aa71 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Wed, 3 Feb 2016 13:35:42 +0530
Subject: [PATCH 13/13] [EM/USBC_PD]: Fix Buffer Overflow Array Index Out of
 Bounds issue

In set request capability, to get the current and voltage value from
the data object, the data object index should be less than the max
data objects. Hence limited the data object index to be within a
limit of maximum data object count.

Change-Id: I7905db7cdde095dbd177c0bd2e2ee2064b2946d9
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-18679
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
Reviewed-on: https://android.intel.com:443/465842
---
 drivers/usb/typec/pd/policy_engine.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/typec/pd/policy_engine.c b/drivers/usb/typec/pd/policy_engine.c
index 2ffff0a..b70d15b 100644
--- a/drivers/usb/typec/pd/policy_engine.c
+++ b/drivers/usb/typec/pd/policy_engine.c
@@ -1045,7 +1045,7 @@ static void pe_send_self_sink_caps(struct policy_engine *pe)
 }
 
 /*
- * This function will pick one to received caps fron source
+ * This function will pick one to received caps from source
  * based on current sysntem required caps given by DPM.
  */
 static int pe_sink_set_request_cap(struct policy_engine *pe,
@@ -1056,11 +1056,14 @@ static int pe_sink_set_request_cap(struct policy_engine *pe,
 	int i;
 	int mv = 0;
 	int ma = 0;
+	int num_data_objs = rcv_pdos->num_pdos;
 	bool is_mv_match = false;
 
 	rcap->cap_mismatch = true;
+	if (num_data_objs > MAX_NUM_DATA_OBJ)
+		num_data_objs = MAX_NUM_DATA_OBJ;
 
-	for (i = 0; i < rcv_pdos->num_pdos; i++) {
+	for (i = 0; i < num_data_objs; i++) {
 		/*
 		 * FIXME: should be selected based on the power (V*I) cap.
 		 */
-- 
1.9.1

